### PR TITLE
add type declarations to C code printer

### LINKED ISSUE #10446: Add type declarations to C code printer.
I find myself prepending the type declaration to C code a lot and think that this should probably be an option to the printer. Something like:

```
In [3]: from sympy import ccode, symbols

In [5]: a, b = symbols('a, b')

In [6]: ccode(a**2, b)
Out[6]: 'b = pow(a, 2);'

In [7]: ccode(a**2, b, declare='double')
Out [7]: 'double b = pow(a, 2);'

In [8]: ccode(a, declare='double')
Out [8]: `double a;`
```

### PR DESCRIPTION
**Pull request by AnishShah**
Fixes #10446

### CODE REVIEW

File: sympy/printing/tests/test_ccode.py [hunk 1]
```diff
@@ -476,3 +476,11 @@ def test_ccode_sign():
...
     expr = sign(cos(x))
     assert ccode(expr) == '(((cos(x)) > 0) - ((cos(x)) < 0))'
+
+
+def test_ccode_declare():
+
+    a, b = symbols("a, b")
+
+    assert ccode(a**2, b, declare="double") == "double b = pow(a, 2);"
+    assert ccode(a, declare="double") == "double a;"
```
Comment (Shekharrajak) [inline sympy/printing/tests/test_ccode.py]: Press enter at end of the line.

**Comment by AnishShah** [comment]
Can anyone review this PR?

**Comment by moorepants** [comment]
@asmeurer @jcrist Do you all think this is a reasonable addition?

File: sympy/printing/ccode.py [hunk 2]
```diff
@@ -298,6 +303,8 @@ def ccode(expr, assign_to=None, **settings):
         Setting contract=False will not generate loops, instead the user is
         responsible to provide values for the indices in the code.
         [default=True].
+    declare: string, optional
+        When given, it specifies the data type of `assign_to` variable.
```
Comment (moorepants) [inline sympy/printing/ccode.py]: The above isn't true for this case: `ccode(a, declare="double")`. Please make it generalize to what it actually does.

**Comment by AnishShah** [comment]
@moorepants I made the changes! :) Let me know if you require any other changes.

**Comment by moorepants** [comment]
I'd like to get @asmeurer feedback as we recently decided not to support summation in C. We may not want to extend this too far away from printing simple expressions.

**Comment by asmeurer** [comment]
There is some typing stuff inherited from symcc at https://github.com/sympy/sympy/pull/10486 (I haven't ported the code printers yet).  I haven't really decided what to do with it, though. Personally type declarations are a lower priority for me. I am afraid of not doing them correctly, so I am a little hesitant on this change.

**Comment by AnishShah** [comment]
No problem @asmeurer . Should I close this PR?

**Comment by asmeurer** [comment]
FYI I removed the datatype stuff from my PR. I'm unclear what the best way to deal with types is in the codegen module, and it's not a top priority for me to fix it out at the moment. 

@moorepants how useful would the change here be for you? I guess the alternative is to keep track of your variables and declare them separately.

**Comment by moorepants** [comment]
Right now I have to specify these like:

```python
'double {}.format(ccode(a**2 + b, assign=z))
```

Or make a template. I suppose I could subclass the C printer too. This is just a convenience, but in C this is such a common thing to do, it seems like it should be available somehow.

**Pull request closed without merging**